tests/kernel/sleep/src/main.c

```diff
@@ -243,7 +243,7 @@ static void forever_thread_entry(void *p1, void *p2, void *p3)
 	int32_t ret;
 
 	ret = k_sleep(K_FOREVER);
```

Comment:
Shouldn't this be akin to k_thread_suspend()? This is quite confusing to see a call that reads as...

"sleep forever"

returning as

"not possible"

when clearly it is with k_thread_suspend

Reply:
There was a time when k_sleep(K_FOREVER) literally converted to a k_thread_suspend(). However, with the introduction of the _THREAD_SLEEPING thread state bit, that has changed. k_sleep(K_FOREVER) will result in suspend-like behavior insofar as it will not be schedulable and nor will it have a timeout, but according to the kernel bookkeeping it will be sleeping. As a result, all threads that go to sleep are awakened via k_wakeup() or when their sleep period expires.

Aside: It might seem weird, but we can now have a thread be both sleeping and suspended. Such a thread will not resume until it has been both awakened and resumed.